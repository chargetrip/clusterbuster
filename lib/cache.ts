import sha1 from 'sha1';
import { TileCacheOptions, TTtl } from './types/index';

/**
 * @description The default options for the cache
 */
export const defaultCacheOptions: TileCacheOptions = {
  enabled: true,
  type: 'lru-cache',
  lruOptions: {
    max: 100000,
    length: function(n, key) {
      return n * 2 + key.length;
    },
    maxAge: 1000 * 60 * 60,
  },
  redisOptions: {
    ttl: 86400, // 24 hours
    host: process.env.REDIS_HOST,
  },
};

export function Cache(
  customCacheOptions: TileCacheOptions = defaultCacheOptions
) {
  let lruCache: any = null;
  let redisCache: any = null;
  const cacheOptions = {
    ...defaultCacheOptions,
    ...customCacheOptions,
    lruOptions: {
      ...defaultCacheOptions.lruOptions,
      ...customCacheOptions.lruOptions,
    },
    redisOptions: {
      ...defaultCacheOptions.redisOptions,
      ...customCacheOptions.redisOptions,
      dropBufferSupport: false,
    },
  };

  if (cacheOptions.enabled) {
    if (cacheOptions.type === 'lru-cache') {
      const LRU = require('lru-cache');
  
      lruCache = new LRU(cacheOptions.lruOptions);
    } else if (cacheOptions.type === 'redis') {
      const Redis = require('ioredis');
  
      redisCache = new Redis(cacheOptions.redisOptions);
    }
  }

  return {
    /**
     * @description Get the cache key for a table, tile data and where statement of the filters
     *
     * @param {string} table The cache table name
     * @param {number} z The tile zoom level
     * @param {number} x The tile x position
     * @param {number} y The tile y position
     * @param {string[]} filters The where statement
     * @returns The cache key
     */
    getCacheKey: (
      table: string,
      z: number,
      x: number,
      y: number,
      filters: string[]
    ): string | null => {
      if (!cacheOptions.enabled) {
        return null;
      }
      const where = sha1(
        filters.sort((a, b) => (a > b ? 1 : a < b ? -1 : 0)).join('-')
      );

      return `${table}-${z}-${x}-${y}-${where}`;
    },

    /**
     * @description Get the cache value of a cache key
     *
     * @param {string} key The cache key
     * @returns The cache value or null if not found or disabled
     */
    getCacheValue: async (key: string): Promise<any> => {
      if (!cacheOptions.enabled) {
        return null;
      }

      if (cacheOptions.type === 'lru-cache') {
        return lruCache.get(key);
      }

      if (cacheOptions.type === 'redis') {
        return await redisCache.getBuffer(key);
      }

      // Invalid type
      return null;
    },
    /**
     * @description Set the cache value for a cache key
     *
     * @param {string} key The cache key
     * @param {any} value The cache value
     * @param {number} zoomLevel The zoom level requested
     * @param {number | TTtl} ttl The time to leave of the cache
     */
    setCacheValue: async (
      key: string,
      value: any,
      ttl?: number
    ): Promise<void> => {
      if (!cacheOptions.enabled) {
        return;
      }

      if (cacheOptions.type === 'lru-cache') {
        lruCache.set(key, value);
      } else if (cacheOptions.type === 'redis') {
        if (!!ttl) {
          await redisCache.set(key, value, 'EX', ttl);
        } else {
          await redisCache.set(key, value);
        }
      }
    },

    /**
     * @description Get the cache TTL from the zoom level and request cache TTL or config cache TTL
     *
     * @param {number} zoomLevel The zoom level requested
     * @param {number | TTtl} ttl The time to leave of the cache
     */
    getCacheTtl: (zoomLevel: number, ttl?: number | TTtl): number => {
      if (
        !cacheOptions.enabled ||
        cacheOptions.type !== 'redis'
      ) {
        return 0;
      }

      const requestTtl = ttl || cacheOptions.redisOptions.ttl;

      if (!!requestTtl) {
        return !isNaN(requestTtl as number)
          ? (requestTtl as number)
          : (requestTtl as TTtl)(zoomLevel);
      }

      return 0;
    },
  };
}
