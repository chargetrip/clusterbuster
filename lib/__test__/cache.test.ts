import { Cache } from '../cache';

jest.mock('ioredis');
jest.mock('lru-cache');

describe('getCacheKey', () => {
  it('should generate key', () => {
    const cache = Cache();
    expect(cache.getCacheKey('table', 0, 1, 2, ['1', '2'])).toMatchSnapshot();
    expect(cache.getCacheKey('table', 0, 1, 2, ['2', '1'])).toMatchSnapshot();
  });
});

describe('getCacheValue', () => {
  it('should get no value', async () => {
    const cache = Cache({
      enabled: false,
      enable: false,
    });
    expect(await cache.getCacheValue('key')).toMatchSnapshot();
  });

  it('should get lru with a value', async () => {
    const cache = Cache({
      enabled: true,
      enable: true,
      type: 'lru-cache',
    });
    expect(await cache.getCacheValue('key')).toMatchSnapshot();
  });

  it('should get redis with a value', async () => {
    const cache = Cache({
      enabled: true,
      enable: true,
      type: 'redis',
    });
    expect(await cache.getCacheValue('key')).toMatchSnapshot();
  });
});

describe('getCacheTtl', () => {
  const cache = Cache({
      enabled: true,
      enable: true,
      type: 'redis',
      redisOptions: {
        ttl: 3600,
      },
    }),
    zoomLevelCache = [3600, 2400, 1800];

  it('should use global ttl', () => {
    expect(cache.getCacheTtl(0)).toMatchSnapshot();
  });

  it('should use request ttl as number', () => {
    expect(cache.getCacheTtl(0, 1800)).toMatchSnapshot();
  });

  it('should use request ttl as function', () => {
    expect(cache.getCacheTtl(0, () => 2400)).toMatchSnapshot();
  });

  it('should use request ttl from array', () => {
    for (const z of [0, 1, 2, 3, 4, 5]) {
      expect(
        cache.getCacheTtl(z, (zoomLevel: number) =>
          typeof zoomLevelCache[zoomLevel] !== 'undefined'
            ? zoomLevelCache[zoomLevel]
            : 0
        )
      ).toMatchSnapshot();
    }
  });
});
