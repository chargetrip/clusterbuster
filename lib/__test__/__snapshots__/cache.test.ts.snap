// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCacheKey should generate key 1`] = `"table-0-1-2-b8a2645298053fb62ea03e27feea6c483d3fd27e"`;

exports[`getCacheKey should generate key 2`] = `"table-0-1-2-b8a2645298053fb62ea03e27feea6c483d3fd27e"`;

exports[`getCacheTtl should use global ttl 1`] = `3600`;

exports[`getCacheTtl should use request ttl as function 1`] = `2400`;

exports[`getCacheTtl should use request ttl as number 1`] = `1800`;

exports[`getCacheTtl should use request ttl from array 1`] = `3600`;

exports[`getCacheTtl should use request ttl from array 2`] = `2400`;

exports[`getCacheTtl should use request ttl from array 3`] = `1800`;

exports[`getCacheTtl should use request ttl from array 4`] = `0`;

exports[`getCacheTtl should use request ttl from array 5`] = `0`;

exports[`getCacheTtl should use request ttl from array 6`] = `0`;

exports[`getCacheValue should get lru with a value 1`] = `
Object {
  "data": Array [],
  "type": "Buffer",
}
`;

exports[`getCacheValue should get no value 1`] = `null`;

exports[`getCacheValue should get redis with a value 1`] = `
Object {
  "data": Array [],
  "type": "Buffer",
}
`;
